### PR TITLE
Rewrite the control module

### PR DESCRIPTION
**Pull request by paultag**
@tianon needs review

### CODE REVIEW

**Comment by paultag** [comment]
I'd like to squash commits; let me know thoughts, and I'll squash at the end

File: control/decode_test.go [hunk 1]
```diff
@@ -37,8 +37,9 @@ Foo-Bar: baz
 
 Value: Bar
 
-Value Baz
+Value: Baz
```
Comment (tianon) [inline control/decode_test.go]: Hahahaha, good catch! :+1:

File: control/parse_test.go [hunk 2]
```diff
@@ -59,119 +59,218 @@ func assert(t *testing.T, expr bool) {
...
+Format: 1.8
+Date: Mon, 16 Nov 2015 21:15:55 -0800
+Source: hy
+Binary: hy python-hy python3-hy
+Architecture: source
+Version: 0.11.0-4
+Distribution: unstable
+Urgency: medium
+Maintainer: Tianon Gravi <tianon@debian.org>
+Changed-By: Tianon Gravi <tianon@debian.org>
```
Comment (tianon) [inline control/parse_test.go]: ANYBODY WANNA GET HYYYY??? :smile: :tada: :balloon:

File: control/parse_test.go [hunk 3]
```diff
@@ -59,119 +59,218 @@ func assert(t *testing.T, expr bool) {
...
-	assert(t, deb822 != nil)
+	notok(t, err)
+}
 
-	assert(t, len(deb822.Order) == 3)
-	assert(t, len(deb822.Order) == len(deb822.Values))
-	assert(t, deb822.Values["Foo"] == "bar")
+func TestLineWrapping(t *testing.T) {
+	// Reader {{{
+	reader, err := control.NewParagraphReader(strings.NewReader(`-----BEGIN PGP SIGNED MESSAGE-----
```
Comment (tianon) [inline control/parse_test.go]: Is this the same example paragraph each time?  Would it make sense to make it a `const`?

**Comment by tianon** [comment]
Feeling pretty honored to have made the upload that is the test case signed paragraph here! :smile:

Overall, I think this is looking really sweet -- like we talked on IRC it needs some unmarshal love WRT signed paragraphs, but this is a really good refactor. :+1:

File: control/decode_test.go (hunk 1 above)
Comment (paultag) [inline control/decode_test.go]: T E S T I N G!

File: control/parse_test.go (hunk 3 above)
Comment (paultag) [inline control/parse_test.go]: Yeah, good idea

**Comment by paultag** [comment]
@stapelberg This is a pretty big changeset; if you've got any code using this, you should make sure it doesn't trigger anything funny (I haven't changed the API for Unmarshal, so it should be fine unless you were using underlying APIs)

File: control/parse_test.go [hunk 4]
```diff
@@ -59,119 +117,108 @@ func assert(t *testing.T, expr bool) {
...
+nl26axBfgoj/kvNLXGOR+rpGz0IBE7QpecVh7eFUb87i4En+lZi9pl0+hC/2n+Xy
+vShNXphZZQ444P5CMX5s8OK0IbWl1wVe0OCjcQhW9juOIGdn3bxWJ68HaIbw5Y+V
+TZcmHxjJJbO+D+ng3OHqCg1tooy1dAeMZlRwsgYyDtx0Rhd7gZKzx5NkuWlBOdSH
+P+FikrKFHW6rvvO0esWqgm7GuBDrMrsPgU9T4UZ1sOOwsBdWTgp9QWceFIrrptX0
+C/hiBXZkJP/cXueZQ38GDny6ahuR5HDmNwHhvV/EuZ28GOdKzCxcOCyhoNI2xgOg
+A8Ija2WnFdScMVRuMxDxK8yMdy1/BtZQKV6uzSt7ebHfPcUopBM4yARM8C90EbJD
+/FevdZ9cGw/0bCyun86t
+=dtCZ
+-----END PGP SIGNATURE-----
+`), &keyring)
```
Comment (tianon) [inline control/parse_test.go]: Isn't this block using that same signed paragraph?

**Comment by stapelberg** [comment]
Not cool :(

```
# github.com/Debian/ratt
./ratt.go:156: undefined: control.ParseParagraph
./ratt.go:214: undefined: control.ParseParagraph
```

It’s customary in the Go community to publish changes which break the API under a new package name, e.g. pault.ag/go/debian/control2.

Would you be willing to revert this API breakage and publish it properly?

**Comment by paultag** [comment]
@stapelberg So, I don't want to make `control2`. The API is largely the same, and the "real" API (using `Unmarshal` to read control files) is untouched.

I get that this sucks, so I'd be up for putting a `ParseParagraph` function back in to emulate the old API, but I'm likely going to put deprecation warnings in it. I'll push that ASAP. Would that work?

I should have done that, but I honestly thought no one uses the `ParseParagraph` API, since it's pretty clumsy

**Comment by stapelberg** [comment]
Having a wrapper function for compatibility works for me.

I’ll award extra bonus points if you want to move ratt to use the “real” API, otherwise I’ll do that when I get around to it :). Thanks!

**Comment by paultag** [comment]
@stapelberg Sent in PR https://github.com/Debian/ratt/pull/5 - I think it's net positive :)

Look OK? I'll see if I can get a wrapper in place for API compat anyway soon.

The whole change need to happen because the modeling as existed was broken for OpenPGP signed data, the new knobs account for that a lot better

**Pull request merged**